### PR TITLE
[Feature]: Add OpenAI server prompt_logprobs support #6508

### LINKED ISSUE #6508: [Feature]: Add OpenAI server `prompt_logprobs` support
### 🚀 The feature, motivation and pitch

As noted in documentation OpenAI API don't support outputing only one token
But it's a very strong advantage over commertial models
Been able to get logits for prompt tokens

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by gnpinkert**
This commit adds a prompt_logprobs option in the extra body field of the chat completions API. When set to true, it will return the log probabilities of the decoded input tokens.

This option was not included in the streaming API. This decision was made since streaming is meant for real time feedback with reduced latency, it doesn't make much sense to include the same prompt log probabilities every single time. This can be included if that is also deemed to be useful.

Currently, the server will report an error if stream and prompt_logprobs are both enabled.

The return value in the chat completions API was modeled after the prompt_logprobs return value during offline inference to reduce coding complexity if switching between online/offline.

It was possible to get the prompt_logprobs earlier if echo and top_logprobs were enabled. This behavior was kept the same to not break any existing configurations.

FIX #6508

### CODE REVIEW

File: vllm/entrypoints/openai/protocol.py [hunk 1]
```diff
@@ -161,6 +162,13 @@ class ChatCompletionRequest(OpenAIBaseModel):
             "If true, the new message will be prepended with the last message "
             "if they belong to the same role."),
     )
+
+    prompt_logprobs: bool = Field(
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/entrypoints/openai/protocol.py:166]: Also, should we allow integer values so it's the same as in `SamplingParams`? (In that case we should move it inside the `chat-completion-sampling-params` block)

**Comment by DarkLight1337** [comment]
It would also be nice if you add this to the Completions API as well (not just Chat Completions API).

File: vllm/entrypoints/openai/serving_chat.py [hunk 2]
```diff
@@ -83,6 +83,18 @@ async def create_chat_completion(
...
+            if request.stream:
+                assert (request.prompt_logprobs is None
+                        or not request.prompt_logprobs), (
+                            "Prompt_logprobs are not "
+                            "available when stream is "
+                            "enabled")
+        except ValueError as e:
+            logger.error(
+                "Mutually exclusive options in chat completion request: %s", e)
+            return self.create_error_response(str(e))
```
[the remark this answers is not in the document]
Comment (gnpinkert) [reply vllm/entrypoints/openai/serving_chat.py:96]: I looked throughout the function and tried to match the throwing pattern. What is the preferred way of handling errors/exceptions in the code?

**Comment by gnpinkert** [comment]
> It would also be nice if you add this to the Completions API as well (not just Chat Completions API).

Sure, I can get on that as well

File: vllm/entrypoints/openai/protocol.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gnpinkert) [reply vllm/entrypoints/openai/protocol.py:166]: Default value of false works for me, makes more sense. 

> Also, should we allow integer values so it's the same as in SamplingParams? (In that case we should move it inside the chat-completion-sampling-params block)

It was a bit odd to me that they were different so I think I would prefer for the API to be as uniform as possible. If you agree, I will make that change

File: vllm/entrypoints/openai/serving_chat.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/entrypoints/openai/serving_chat.py:96]: I mean that you can simply log and return the error response without performing assertion just to catch it again.

```py
if request.stream:
    if not (request.prompt_logprobs is None
            or not request.prompt_logprobs):
        return self.create_error_response(
            "Prompt_logprobs are not available when stream is enabled")
```

**Comment by gnpinkert** [comment]
@DarkLight1337 I have implemented the requested changes in the fixup. I did keep the default value in the response object as 'None' since it appears to be the default value for all unused/unfilled response values. Otherwise, I put the default as 0, as per your suggestion to not default to None.

However, the SamplingParams object does default to None. It would make more sense to also change the default value in SamplingParams to keep consistency.

**Comment by DarkLight1337** [comment]
> @DarkLight1337 I have implemented the requested changes in the fixup. I did keep the default value in the response object as 'None' since it appears to be the default value for all unused/unfilled response values. Otherwise, I put the default as 0, as per your suggestion to not default to None.
> 
> However, the SamplingParams object does default to None. It would make more sense to also change the default value in SamplingParams to keep consistency.

I meant that the default should not be `None` when the value would otherwise only be a boolean. Now that it can be an integer, it is fine to default to `None` again.

File: vllm/entrypoints/openai/protocol.py [hunk 3]
```diff
@@ -532,6 +537,8 @@ class CompletionResponseChoice(OpenAIBaseModel):
             "to stop, None if the completion finished for some other reason "
             "including encountering the EOS token"),
     )
+    prompt_logprobs: Optional[List[Optional[Dict[int, Logprob]]]] = Field(
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/protocol.py:540]: This seems to be a repeated definition.

File: vllm/entrypoints/openai/protocol.py [hunk 4]
```diff
@@ -532,6 +537,8 @@ class CompletionResponseChoice(OpenAIBaseModel):
...
+        default=None)
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/protocol.py:541]:
```suggestion
    prompt_logprobs: Optional[List[Optional[Dict[int, Logprob]]]] = None
```

Using `Field` is unnecessary here.

File: vllm/entrypoints/openai/protocol.py [hunk 5]
```diff
@@ -627,6 +634,8 @@ class ChatCompletionResponse(OpenAIBaseModel):
     model: str
     choices: List[ChatCompletionResponseChoice]
     usage: UsageInfo
+    prompt_logprobs: Optional[List[Optional[Dict[int, Logprob]]]] = Field(
+        default=None)
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/protocol.py:638]: Ditto

File: vllm/entrypoints/openai/serving_chat.py [hunk 6]
```diff
@@ -506,7 +516,8 @@ async def chat_completion_full_generator(
             model=model_name,
             choices=choices,
             usage=usage,
-        )
+            prompt_logprobs=final_res.prompt_logprobs
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/serving_chat.py:519]: Why not just pass `final_res.prompt_logprobs` directly?

Comment (gnpinkert) [reply vllm/entrypoints/openai/serving_chat.py]: good catch, artifact from before the default was back to None

**Review by DarkLight1337 — APPROVED** [review]
LGTM, thanks for implementing this!

File: vllm/entrypoints/openai/serving_chat.py [hunk 7]
```diff
@@ -506,7 +516,7 @@ async def chat_completion_full_generator(
             model=model_name,
             choices=choices,
             usage=usage,
-        )
+            prompt_logprobs=final_res.prompt_logprobs)
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/serving_chat.py:519]:
```suggestion
            prompt_logprobs=final_res.prompt_logprobs,
        )
```

File: vllm/entrypoints/openai/serving_completion.py [hunk 8]
```diff
@@ -377,7 +386,7 @@ def request_output_to_completion_response(
                     logprobs=logprobs,
                     finish_reason=output.finish_reason,
                     stop_reason=output.stop_reason,
-                )
+                    prompt_logprobs=final_res.prompt_logprobs)
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/serving_completion.py:389]:
```suggestion
                    prompt_logprobs=final_res.prompt_logprobs,
                )
```

**Comment by SrGonao** [comment]
Is the correct way to make the request to a vllm server to use ` kwargs["extra_body"] =  {"prompt_logprobs": 15}`? I haven't managed to generate a request that returns the prompt logprobs using the chat completion api. I looked at the tests but this is what they suggest and it does not seem to work in my end. Using the most recent version of vllm

**Comment by DarkLight1337** [comment]
> Is the correct way to make the request to a vllm server to use ` kwargs["extra_body"] = {"prompt_logprobs": 15}`? I haven't managed to generate a request that returns the prompt logprobs using the chat completion api. I looked at the tests but this is what they suggest and it does not seem to work in my end. Using the most recent version of vllm

Yes, please check the test cases for some examples.

**Comment by SrGonao** [comment]
Thank you. I did find it (I was doing it wrong on my end)
